mcs/class/corlib/System.Runtime.InteropServices/Marshal.cs

```diff
@@ -1897,5 +1897,84 @@ internal static void SetLastWin32Error (int error)
 			throw new NotSupportedException();
 		}
 #endif
+
+		internal struct MarshalerInstanceKey {
+			public Type Type;
+			public string Cookie;
+
+			public override int GetHashCode () {
+				int result = 0;
+				if (Type != null)
+					result = Type.GetHashCode();
+				if (Cookie != null)
+					result ^= Cookie.GetHashCode();
+				return result;
+			}
+		}
+
+		internal class MarshalerInstanceKeyComparer : IEqualityComparer<MarshalerInstanceKey> {
+			public bool Equals (MarshalerInstanceKey lhs, MarshalerInstanceKey rhs) {
+				return (lhs.Type == rhs.Type) && (lhs.Cookie == rhs.Cookie);
+			}
+
+			public int GetHashCode (MarshalerInstanceKey key) {
+				return key.GetHashCode ();
+			}
+		}
+
+		internal static Dictionary<MarshalerInstanceKey, ICustomMarshaler> MarshalerInstanceCache = 
+			new Dictionary<MarshalerInstanceKey, ICustomMarshaler> (new MarshalerInstanceKeyComparer ());
+
+		internal static ICustomMarshaler GetCustomMarshalerInstance (Type type, string cookie) {
+			var key = new MarshalerInstanceKey {
+				Type = type,
+				Cookie = cookie
+			};
+
+			ICustomMarshaler result;
+			bool gotExistingInstance;
+			lock (MarshalerInstanceCache)
+				gotExistingInstance = MarshalerInstanceCache.TryGetValue (key, out result);
+
+			if (!gotExistingInstance) {
+				MonoMethod getInstanceMethod;
+				try {
+					getInstanceMethod = (MonoMethod)type.GetMethod (
+						"GetInstance", BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.InvokeMethod,
```

Comment:
Please add test that all this work with linker enabled to mcs/tools/linker/Tests

Reply:
I added linker tests but the linker appears to eat custom marshalers entirely. I think this means we need to fix the linker. For now I guess I should whitelist the marshaler classes?

Reply:
Whitelist Marshall class is too aggressive. I can look at the linker tweaks so just add the test as disabled